### ISSUE TITLE
request.user type is not adjusted following conditions

### ISSUE THREAD

**Issue by AdrienLemaire** [typeddjango/django-stubs#500]
```py
class ManagerProfilePagesMenuItem(SubmenuMenuItem):
    def is_shown(self, request: HttpRequest) -> bool:
        user = request.user
        if user.is_anonymous: return False
        return (
            user.has_perm("userprofile.view_profilepage")
            and user.companies.count() > 1
        )
```

![](https://i.imgur.com/d1XTYAz.png)

Following the line `if user.is_anonymous: return False`, there cannot be any mistake that user is a `User` and not `Union[User, AnonymousUser]` (well actually, `user.has_perm("userprofile.view_profilepage")` should be enough to validate it).
We shouldn't get a type warning for the following `user.companies.count()` condition, since AnonymousUser can never reach this line.

**Comment by sobolevn** [comment]
Hi! Thanks a lot for the bug report. I am pretty sure it was discussed already somewhere, but I cannot find where.

So, here's the deal. We are using `Literal[True]` in our anonymous user. And we are using `Literal[False]` in logger in user: https://github.com/typeddjango/django-stubs/blob/master/django-stubs/contrib/auth/base_user.pyi#L29 and https://github.com/typeddjango/django-stubs/blob/master/django-stubs/contrib/auth/models.pyi#L114

Please, check what type is revealed for `YourUser.is_anonymous`
Any contributions (including tests) are very welcome!

**Comment by flaeppe** [comment]
Closed via: #309 

We have tests covering this:

https://github.com/typeddjango/django-stubs/blob/076e637f7e4d7e9f14375c08911d447eb11def86/tests/typecheck/test_request.yml#L18-L27

Feel free to reopen with a breaking case
